test/acceptance/pushPullTestDataFeeder.scala

```diff
@@ -38,43 +38,67 @@ trait PushPullDBInsertionTestDataFeeder extends Matchers with Eventually {
 
   def makeAPICall: (ExpectedCall => Unit)
 
-  def insertIntoDB: (ExpectedCall => Unit)
-
   def apiSubscriptionFieldsService: ApiSubscriptionFieldsService
 
   def insertTestData(totalNotificationsToBeSent: Int,
-                     numberOfClientsToTest: Int
-                    ): (Map[Client, List[ExpectedCall]], Map[Client, List[ExpectedCall]]) = {
+                     numberOfClientsToTest: Int): (Map[Client, List[ExpectedCall]], Map[Client, List[ExpectedCall]]) = {
 
-    val expectedPushNotificationsByCSID: mutable.Map[Client, ListBuffer[ExpectedCall]] = mutable.Map()
-    val expectedPullNotificationsByCSID: mutable.Map[Client, ListBuffer[ExpectedCall]] = mutable.Map()
+    val expectedPushNotificationsByClient: mutable.Map[Client, ListBuffer[ExpectedCall]] = mutable.Map()
+    val expectedPullNotificationsByClient: mutable.Map[Client, ListBuffer[ExpectedCall]] = mutable.Map()
 
-    val poolOfPushOrPullEnabledClientCSIDs: List[Client] = createPoolOfPushEnabledDisabledClients(numberOfClientsToTest, mutable.Set())(apiSubscriptionFieldsService).toList
+    val poolOfPushOrPullEnabledClients: List[Client] = createPoolOfPushEnabledDisabledClients(numberOfClientsToTest, mutable.Set())(apiSubscriptionFieldsService).toList
 
-    for (a <- 1 to totalNotificationsToBeSent) {
+    createTestData(totalNotificationsToBeSent - 1,
+      sendNotificationToCustomsNotificationFor(aRandomlySelectedClient(poolOfPushOrPullEnabledClients)()),
+      expectedPushNotificationsByClient,
+      expectedPullNotificationsByClient,
+      aRandomlySelectedClient(poolOfPushOrPullEnabledClients) _)
 
-      val client = poolOfPushOrPullEnabledClientCSIDs(Random.nextInt(numberOfClientsToTest))
 
-      val insertNotificationInDB = false //Random.nextBoolean()
-      val processor: ExpectedCall => Unit = if (insertNotificationInDB) insertIntoDB else makeAPICall
+    //Wait until all requests are made i.e. all above Futures are completed.
+    //If this starts failing then increase the PatienceConfig.
+    eventually(expectedPushNotificationsByClient.values.flatten.size + expectedPullNotificationsByClient.values.flatten.size should be(totalNotificationsToBeSent))
 
-      if (client.isPushEnabled) {
-        fireOffRandomRequest(client, processor).map { expectedCall =>
-          addExpectedCallTo(expectedPushNotificationsByCSID, expectedCall)
-        }
-      } else {
-        fireOffRandomRequest(client, processor).map { expectedCall =>
-          addExpectedCallTo(expectedPullNotificationsByCSID, expectedCall)
-        }
+    (convertToMutableMap(expectedPushNotificationsByClient), convertToMutableMap(expectedPullNotificationsByClient))
+  }
+
+  private def aRandomlySelectedClient(poolOfPushOrPullEnabledClientCSIDs: List[Client])() =
+    poolOfPushOrPullEnabledClientCSIDs(Random.nextInt(poolOfPushOrPullEnabledClientCSIDs.size))
+
```

Comment:
what do you mean by "Create an Seq of random clients, of range1 to totalNotificationsToBeSent", clients and totalNotificationsToBeSent are two different things.

Reply:
If i understood you correctly then this function should return List[Future[ExpectedCall]]. but if I pass a seq of Futures to this method then the Futures are not in order or the execution cant be guaranteed. Sorry, I am getting how you think it fits in this situation. Plus I think this createTestData is simpler then this sequence.

Reply:
No you pass in a Seq of randomly selected Clients from the pool 1 to totalNotificationsToBeSent

Reply:
I think I got you now, however, I think that passing in a list randomly selected clients when we actually need a list randomly generated requests is shoehorning to use this method. It would be

Reply:
technically both are same thing, readability wise, it is your call. I still find it confusing that we'll have to pass in a list randomly selected client when we actually need list of randomly generated calls.

Reply:
👍